contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,609 @@
+            price = type(uint256).max / (price > 0 ? price : 1);
+        }
+
+        return price;
+    }
+
+    // =================================== ROLLOVER VALIDATION =========================================
+    /**
+     * @notice Validates that the rollover is valid. If any of these conditionals are not met
+     *         the transaction will revert.
+     *
+     * @dev All whitelisted payable currencies and collateral must be whitelisted.
+     *
+     * @param sourceLoanTerms           The terms of the original loan.
+     * @param newLoanTerms              The terms of the new loan.
+     * @param borrowerNoteId            The ID of the borrowerNote for the old loan.
+     */
+    function _validateCurrencyRollover(
+        LoanLibrary.LoanTerms memory sourceLoanTerms,
+        LoanLibrary.LoanTerms memory newLoanTerms,
+        uint256 borrowerNoteId
+    ) internal view {
+        // ------------- Caller Validation -------------
+        address _borrower = IPromissoryNote(borrowerNote).ownerOf(borrowerNoteId);
+
+        if (_borrower != msg.sender) revert CCR_CallerNotBorrower();
+
+        // ------------- Rollover Terms Validation -------------
+        // currency must not be the same
+        if (sourceLoanTerms.payableCurrency == newLoanTerms.payableCurrency) {
+            revert CCR_SameCurrency(sourceLoanTerms.payableCurrency, newLoanTerms.payableCurrency);
+        }
+
+        // collateral address and id must be the same
+        if (sourceLoanTerms.collateralAddress != newLoanTerms.collateralAddress || sourceLoanTerms.collateralId != newLoanTerms.collateralId) {
+            revert CCR_CollateralMismatch(
+                sourceLoanTerms.collateralAddress,
+                sourceLoanTerms.collateralId,
+                newLoanTerms.collateralAddress,
+                newLoanTerms.collateralId
+            );
+        }
+
+        // ------------- New LoanTerms Validation -------------
+        originationHelpers.validateLoanTerms(newLoanTerms);
+    }
+
+    // ========================================= HELPERS ================================================
+    /**
+     * @notice Helper function to distribute funds based on the rollover amounts.
+     *
+     * @param oldLoanId                 The ID of the original loan to be rolled over.
+     * @param oldLoanData               The loan data of the original loan.
+     * @param newPrincipalAmount        The principal amount of the new loan.
+     * @param newCurrency               The currency of the new loan.
+     * @param borrower_                 The address of the borrower.
+     * @param lender                    The address of the new lender.
+     *
+     * @return amounts                  The rollover amounts.
+     * @return feeSnapshot              A snapshot of current lending fees.
+     * @return repayAmount              The amount needed to repay the old loan.
+     */
+    function _rollover(
```

Comment:
Given the new flow, think about function naming here; this name was a bit confusing to me since it doesn't actually _do_ the rollover (unlike, for instance, `_rollover` in `OriginationController`, which calls `loanCore.rollover`).

You have three functions which all do different rollover "steps"; `_rollover`, `_initiateRepayment`, and `_initializeRolloverLoan`. Then, within `_initiateRepayment`, you have `_processRepaymentSwap`. 

I'm not really understanding the design choices between breaking the functions down in this way; none of the 4 ever get re-used, and the naming is confusing for sure. Like, `_initiateRepayment` _doesn't_ repay the loan, `_processRepaymentSwap` does, but on first reading, that name would seem to suggest that it's only handling the swap (in fact, it's doing the swap, _and_, the repay, _and_ settling funds with the borrower, via _another_ internal function, `_repayLoan`).

At the deepest, you're getting _four_ internal functions deep in the callstack:

1. `rolloverCrossCurrencyLoan` calls:
2. `_initiateRepayment` calls:
3. `_processRepaymentSwap calls:
4. `_repayLoan`.

Moreover, the name-prefix conventions are unclear? What's the difference in meaning between `processXXX` and `initiateXXX`, vs. functions with no prefix at all (like `_rollover` or `_repayLoan`)?

To simplify, the encapsulating function, `rolloverCrossCurrencyLoan`, has to do the following things starting on line 133:

1. Calculate settlement amounts (currently happens in `_rollover`, which name doesn't reflect that`
2. _Perform_ settlement (which is right now broken up across `_rollover` and `_processRepaymentSwap` which is inside `_initiateRepayment`)
3. Repay a loan (which is inside `_repayLoan` which is inside `_processRepaymentSwap` which is inside `_initiateRepayment`)
4. Start a new loan which is inside (`_intializeRolloverLoan`).

Step 4 is pretty clearly marked and delineated logic wise, but steps 1-3 are all a bit jumbled and confusingly-named. I recommend taking another pass at coded organization for these steps, trying to make functions as clearly named and as single-responsibility as possible, and avoiding deep cascading callstacks.